tlog-tiles.md

```diff
@@ -0,0 +1,163 @@
+
+### Merkle Tree
+
+Instead of serving consistency and inclusion proofs for arbitrary entries and/or
+tree sizes, logs serve the Merkle Tree as a set of “tiles”: concatenated
+sequences of consecutive Merkle Tree Hashes at a certain tree height. Clients
+can fetch all the tiles they need in parallel and compute any desired proof.
+
+Tiles are served at
+
+	<prefix>/tile/<L>/<N>[.p/<W>]
+
+with `Content-Type: application/data`.
+
+`<L>` is the “level” of the tile, and MUST be a decimal ASCII integer between 0
+and 63, with no additional leading zeroes.
+
+`<N>` is the index of the tile within the level. It MUST be a non-negative
+integer encoded into zero-padded 3-digit path elements. All but the last path
+element MUST begin with an `x`. For example, index 1234067 will be encoded as
+`x001/x234/067`. (This allows storing tile resources efficiently in a filesystem
+without file/directory conflicts, and serving them directly.)
+
+The `.p/<W>` suffix is only present for partial tiles, defined below. `<W>` is
+the width of the tile, a decimal ASCII integer between 1 and 255, with no
+additional leading zeroes.
+
+This endpoint is immutable, so its caching headers SHOULD be long-lived.
+
+Full tiles MUST be exactly 256 hashes wide, or 8,192 bytes.  At “level 0” tiles
+contain leaf hashes. At “level 1” and above, each hash in a tile is the Merkle
+Tree Hash of a *full* tile at the level below.
+
+More formally, in the language of RFC 6962, Section 2.1, the *n*-th tile at
+level *l*, with *n* and *l* starting at 0, is the sequence of the following
+Merkle Tree Hashes, with *i* from 0 to 255:
+
+    MTH(D[(n * 256 + i) * 256**l : (n * 256 + i + 1) * 256**l])
+
+Note that a tile represents the entire subtree of height 8 with its hashes as
+the leaves. The Merkle Tree levels between those expressed by the tile hashes
+are reconstructed by hashing the leaves.
+
+#### Partial Tiles
+
+Some or all of the rightmost tiles in a tree will always be _partial_. A partial
+tile contains between 1 and 255 hashes, and MUST NOT be hashed into a tile at
+the level above.
+
+The partial tile at level *l* for a tree of size *s* has `floor(s / 256**l) mod
+256` entries. Empty tiles MUST NOT be served.
+
+For example, a tree of size 70,000 will be represented by 273 full level 0
+tiles, one partial level 0 tile of width 112, one full level 1 tile, one partial
+level 1 tile of width 17, and one partial level 2 tile of width 1. Note that a
+tree of size 256 will be represented by a full level 0 tile and a partial level
+1 tile of width 1.
+
+Logs MUST serve partial tiles corresponding to tree sizes for which a checkpoint
+was produced, but MAY delete any partial tile once the corresponding full tile
+becomes available. Clients MUST NOT fetch arbitrary partial tiles without
+verifying a checkpoint with a size that requires their existence, and MAY fetch
+the full tile in parallel as a fallback in case the partial tile is not
+available anymore.
```

Comment:
It would be good to specify an HTTP status code for deleted / non-existent tiles so that clients can distinguish between a successful request, a deleted tile, and an error that might be fixed by retrying.

For example: "Logs MUST return an HTTP status code between 400 and 499 inclusive if a tile does not exist."

Reply:
I want to make sure we are compatible with most file storage systems such as S3, but I guess those will always do 200 for success and 404 or 410 for not found?

Reply:
Ok, looking at https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#ErrorCodeList it seems that some 4xx might be retryable but not 404.

Reply:
I wanted to specify 404 or 410 but Let's Encrypt Twig (which appears to be S3) is returning 403 for non-existent files so I made it 4xx to be safe.

The fact that some 4XX codes are retryable is annoying. 429 (Too Many Requests) is generally retryable as well (though S3 uses 503 for that case!).

Reply:
It sounds like the details of what is retryable is unfortunately backend-specific, so we can't help clients much from the spec. If that's not right, please open an issue and we can add more text to the draft after merging.